fe/src/main/java/org/apache/doris/clone/TabletScanner.java

```diff
@@ -0,0 +1,181 @@
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.clone;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Sets;
+
+import org.apache.doris.catalog.Catalog;
+import org.apache.doris.catalog.Database;
+import org.apache.doris.catalog.MaterializedIndex;
+import org.apache.doris.catalog.OlapTable;
+import org.apache.doris.catalog.Partition;
+import org.apache.doris.catalog.Table;
+import org.apache.doris.catalog.Tablet;
+import org.apache.doris.catalog.Tablet.TabletStatus;
+import org.apache.doris.common.Pair;
+import org.apache.doris.common.util.Daemon;
+import org.apache.doris.system.SystemInfoService;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.List;
+import java.util.Set;
+
+/*
+ * This scanner is responsible for scanning all unhealthy tablets.
+ * It does not responsible for any scheduler of tablet repairing or balance
+ */
+public class TabletScanner extends Daemon {
+    private static final Logger LOG = LogManager.getLogger(TabletScanner.class);
+
+    private static long INTERVAL_MS = 60 * 1000L; // 1min
+
+    private Catalog catalog;
+    private SystemInfoService infoService;
+    private TabletFactory tabletFactory;
+
+    // db id -> (tbl id -> partition id)
+    // priority of replicas of partitions in this table will be set to VERY_HIGH if not healthy
+    private com.google.common.collect.Table<Long, Long, Set<Long>> priors = HashBasedTable.create();
+
+    public TabletScanner() {
+        super("tablet scanner", INTERVAL_MS);
+        catalog = Catalog.getCurrentCatalog();
+        infoService = Catalog.getCurrentSystemInfo();
+        tabletFactory = new TabletFactory();
+    }
+
+    public void addPriors(long dbId, long tblId, List<Long> partitionIds) {
+        Preconditions.checkArgument(!partitionIds.isEmpty());
+        synchronized (priors) {
+            Set<Long> parts = priors.get(dbId, tblId);
+            if (parts == null) {
+                parts = Sets.newHashSet();
+                priors.put(dbId, tblId, parts);
+            }
+            parts.addAll(partitionIds);
+        }
+    }
+
+    /*
+     * For each cycle, TabletScanner will scan all OlapTable's tablet.
+     * If a tablet is not healthy, a TabletInfo will be created and sent to TabletFactory for repairing.
+     */
+    @Override
+    protected void runOneCycle() {
+
+        if (!tabletFactory.isEmpty()) {
+            LOG.info("tablet factory has unfinished tasks. skip this round of checking");
+            return;
+        }
+
+        scanTablets();
```

Comment:
The exception need catching in `runOneCycle`

Reply:
Here we only catch expected exception, any unexpected exception is caught in `Deamon`